dev/udf_proxy.h

```diff
@@ -0,0 +1,124 @@
+            using final_call_fn_t = void (*)(void* udfHandle, sqlite3_context* context);
+
+            struct destruct_only_deleter {
+                template<class F>
+                void operator()(F* f) const noexcept {
+                    f->~F();
+                }
+            };
+
+            std::string name;
+            int argumentsCount;
+            std::function<void*(void* location)> constructAt;
+            xdestroy_fn_t destroy;
+            func_call_fn_t func;
+            final_call_fn_t finalAggregateCall;
+            // flag whether the UDF has been constructed at `udfHandle`;
+            // necessary for aggregation operations
+            bool constructed;
+            // pointer to memory for UDF in derived proxy veneer
+            void* const udfHandle;
+        };
+
+        /*
+         *  A veneer to `udf_proxy` that provides memory space for a user-defined function object.
+         *  
+         *  Note: it must be a veneer, i.e. w/o any non-trivially destructible member variables.
+         */
+        template<class UDF>
+        struct udf_proxy_veneer : udf_proxy {
+            // allocated memory for user-defined function
+            alignas(UDF) char storage[sizeof(UDF)];
+
+            udf_proxy_veneer(std::string name,
+                             int argumentsCount,
+                             std::function<void*(void* location)> constructAt,
+                             xdestroy_fn_t destroy,
+                             func_call_fn_t run) :
+                udf_proxy {
+                std::move(name), argumentsCount, std::move(constructAt), destroy, run, nullptr, false,
+#if __cpp_lib_launder >= 201606L
+                    std::launder((UDF*)this->storage)
+#else
+                    storage
+#endif
+            }
+            {}
+
+            udf_proxy_veneer(std::string name,
+                             int argumentsCount,
+                             std::function<void*(void* location)> constructAt,
+                             xdestroy_fn_t destroy,
+                             func_call_fn_t step,
+                             final_call_fn_t finalCall) :
+                udf_proxy {
+                std::move(name), argumentsCount, std::move(constructAt), destroy, step, finalCall, false,
+#if __cpp_lib_launder >= 201606L
+                    std::launder((UDF*)this->storage)
+#else
+                    storage
+#endif
+            }
+            {}
+        };
+
+        inline void scalar_function_callback(sqlite3_context* context, int argsCount, sqlite3_value** values) {
+            udf_proxy* proxy = static_cast<udf_proxy*>(sqlite3_user_data(context));
+            if(proxy->argumentsCount != -1 && proxy->argumentsCount != argsCount) {
+                throw std::system_error{orm_error_code::arguments_count_does_not_match};
+            }
+            const std::unique_ptr<void, xdestroy_fn_t> udfHandle{proxy->constructAt(proxy->udfHandle), proxy->destroy};

```

Comment:
what if this callback is called from several threads simultaneously? Looks like it will produce data races cause we have only one pack of bytes for function object but several callers. Current `dev` implementation doesn't have this issue

Reply:
Your concern is absolutely valid, and I have dealt with this issue of user-defined functions (aka "application-defined SQL functions") myself.

There are 2 things to consider:

1. if you think about it, you come to the conclusion that if SQLite were to impose thread-safety on a callback, it would be an almost impossible requirement to fulfill. This would mean that sqlite_orm would have to document that a user-defined function must be programmed to be thread-safe (we never know what the user-defined function is doing). Also, the SQLite documentation does not mention that a callback must be inherently thread-safe.
1. the callback of the user-defined function is part of a database connection. Therefore, there is a separate "proxy" object for each connection.
  Also, I checked the callstack, and the call to a user-defined function is part of evaluating a prepared statement, i.e. the generation of result rows with `sqlite3_step()`. The `sqlite3_step()` function in turn locks a database mutex, as you can see in the attached screenshot of the callstack. From another perspective, the step function itself would not create multiple threads, but multiple threads could be created to call the step function.

The conclusion is that calling a user-defined function as part of a step during the generation of result rows is inherently concurrency-safe.

![image](https://github.com/fnc12/sqlite_orm/assets/5430212/11746f16-1e02-458e-b47e-c9d74b50105e)